tokio/src/task/task_set.rs

```diff
@@ -0,0 +1,224 @@
+    ///
+    /// Returns `None` if the set is empty.
+    ///
+    /// # Cancel Safety
+    ///
+    /// This method is cancel safe. If `join_one` is used as the event in a `tokio::select!`
+    /// statement and some other branch completes first, it is guaranteed that no tasks were
+    /// removed from this `TaskSet`.
+    pub async fn join_one(&mut self) -> Result<Option<T>, JoinError> {
+        crate::future::poll_fn(|cx| self.poll_join_one(cx)).await
+    }
+
+    /// Poll for one of the tasks in the set to complete.
+    ///
+    /// If this returns `Poll::Ready(Some(_))`, then the task that completed is removed from the
+    /// set.
+    ///
+    /// When the method returns `Poll::Pending`, the `Waker` in the provided `Context` is scheduled
+    /// to receive a wakeup when a task in the `TaskSet` completes. Note that on multiple calls to
+    /// `poll_join_one`, only the `Waker` from the `Context` passed to the most recent call is
+    /// scheduled to receive a wakeup.
+    ///
+    /// # Return value
+    ///
+    /// This function returns:
+    ///
+    ///  * `Poll::Pending` if the `TaskSet` is not empty and all tasks are currently running.
+    ///  * `Poll::Ready(Some(Ok(value)))` if one of the tasks in this `TaskSet` has completed. The
+    ///    `value` is the return value of one of the tasks that completed.
+    ///  * `Poll::Ready(Some(Err(err)))` if one of the tasks in this `TaskSet` has panicked or been
+    ///     aborted.
+    ///  * `Poll::Ready(None)` if the `TaskSet` is empty.
+    pub fn poll_join_one(&mut self, cx: &mut Context<'_>) -> Poll<Result<Option<T>, JoinError>> {
+        loop {
+            // The call to `pop_notified` moves the entry to the `idle` list. It is moved back to
+            // the `notified` list if the waker is notified in the `poll` call below.
+            let entry = match self.inner.pop_notified(cx.waker()) {
+                Some(entry) => entry,
+                None => {
+                    if self.is_empty() {
+                        return Poll::Ready(Ok(None));
+                    } else {
+                        // The waker was set by `pop_notified`.
+                        return Poll::Pending;
+                    }
+                }
+            };
+
+            let res = self.inner.with_entry_value(&entry, |jh| {
+                jh.as_mut()
+                    .map(|jh| entry.with_context(|ctx| Pin::new(jh).poll(ctx)))
+            });
+
+            match res {
+                Some(Poll::Ready(Ok(res))) => {
+                    self.inner.with_entry_value(&entry, |jh| jh.take());
+                    self.inner.remove(&entry);
+                    return Poll::Ready(Ok(Some(res)));
+                }
+                Some(Poll::Ready(Err(err))) => {
+                    self.inner.with_entry_value(&entry, |jh| jh.take());
+                    self.inner.remove(&entry);
+                    return Poll::Ready(Err(err));
+                }
```

Comment:
Looks like the code could be deduped here. I don't *think* we need to re-map it to `Result<Option<T>, ...>` `Option<Result<...>>` should be fine?